engine/src/main/java/org/camunda/bpm/engine/impl/cmd/batch/variables/SetVariablesToProcessInstancesBatchCmd.java

```diff
@@ -0,0 +1,170 @@
+    ensureAtLeastOneNotNull(
+        "No process instances found.",
+        processInstanceIds,
+        processInstanceQuery,
+        historicProcessInstanceQuery
+    );
+
+    BatchElementConfiguration elementConfiguration = collectProcessInstanceIds(commandContext);
+
+    List<String> ids = elementConfiguration.getIds();
+    ensureNotEmpty(BadUserRequestException.class, "processInstanceIds", ids);
+
+    BatchConfiguration configuration = getConfiguration(elementConfiguration);
+    Batch batch = new BatchBuilder(commandContext)
+        .type(Batch.TYPE_SET_VARIABLES)
+        .config(configuration)
+        .permission(BatchPermissions.CREATE_BATCH_SET_VARIABLES)
+        .operationLogHandler(this::writeUserOperationLog)
+        .build();
+
+    String batchId = batch.getId();
+    VariableUtil.setVariables(variables, (name, value) -> setVariable(batchId, name, value));
+
+    return batch;
+  }
+
+  protected void writeUserOperationLog(CommandContext commandContext, int instancesCount) {
+    List<PropertyChange> propChanges = new ArrayList<>();
+
+    int variablesCount = variables.size();
+    propChanges.add(new PropertyChange("nrOfInstances", null, instancesCount));
+    propChanges.add(new PropertyChange("nrOfVariables", null, variablesCount));
+    propChanges.add(new PropertyChange("async", null, true));
+
+    commandContext.getOperationLogManager()
+        .logBatchOperation(UserOperationLogEntry.OPERATION_TYPE_SET_VARIABLES_BATCH, propChanges);
+  }
+
+  protected void setVariable(String batchId, String variableName, Object variableValue) {
+    TypedValue variableTypedValue = Variables.untypedValue(variableValue);
+
+    boolean isTransient = variableTypedValue.isTransient();
+    if (isTransient) {
+      throw LOGGER.exceptionSettingTransientVariablesAsyncNotSupported(variableName);
+    }
+
+    VariableUtil.checkJavaSerialization(variableName, variableTypedValue);
+
+    VariableInstanceEntity variableInstance =
+        VariableInstanceEntity.createAndInsert(variableName, variableTypedValue);
+
+    variableInstance.setVariableScopeId(batchId);
+  }
+
+  public BatchConfiguration getConfiguration(BatchElementConfiguration elementConfiguration) {
+    DeploymentMappings mappings = elementConfiguration.getMappings();
+    List<String> ids = elementConfiguration.getIds();
+
+    return new BatchConfiguration(ids, mappings);
+  }
+
+
+  protected BatchElementConfiguration collectProcessInstanceIds(CommandContext commandContext) {
+    BatchElementConfiguration elementConfiguration = new BatchElementConfiguration();
+
+    if (!CollectionUtil.isEmpty(processInstanceIds)) {
+      ProcessInstanceQueryImpl query = new ProcessInstanceQueryImpl();
+      query.processInstanceIds(new HashSet<>(processInstanceIds));
+      List<ImmutablePair<String, String>> mappings =
+          commandContext.runWithoutAuthorization(query::listDeploymentIdMappings);
+      elementConfiguration.addDeploymentMappings(mappings);
```

Comment:
It might be useful to use this here, in case there are instances no deployment can be found for
```suggestion
      elementConfiguration.addDeploymentMappings(mappings, processInstanceIds);
```

Reply:
We don't need this here since a running process instance cannot exist without a deployment.